docs/rfcs/099-safekeeper-s3-integration.md

```diff
@@ -0,0 +1,86 @@
+# What
+
+Safe keeper needs to backup WAL to S3
+
+S3 is used across the Neon for a variety of components and sharing code will reduce code maintanance cost 
+
+# Why
+
+In situations when other sources of data are compromised transaction log (WAL) produced by compute becomes the ultimate recovery solution 
+
+see https://github.com/zenithdb/zenith/issues/1169 
+
+
+# Proposal
+
+
+## A common S3 integration
+
+Given that S3 integration is common between both Safekeepers and Page Servers a common library can be built
+The common requirement is an ability for asynchrounous upload of objects to a Remote Storage.
+
+Objects can be represented by a byte stream or by a file, optional compression can be applied on either
+Remote storage shuold have production implementation and test implementations. Production impementation is built to work with S3 API compatible blob storage while test implementation is based on a filesystem
+
+ Commmon operations to be supported:
+  1. Upload object (inputs: an object and a destination)
+  2. List objects (inputs: an object filter)  
+  *this is irrelevant to Safekeepers and no discussed here*
+  3. Download object (inputs: a destination)  
+  *this is irrelevant to Safekeepers and no discussed here*
```

Comment:
Can it be useful to dynamically load WAL segments from S3? Right now we say that we'll delete WAL from disk once it's uploaded to S3 and that will break replication from older LSNs. We can try to fix it by dynamically streaming WAL from S3 when there is no longer file on disk.

Reply:
it could be, in fact, it would reduce disk footprint when RO are lagging. I'd keep it out of scope for this document though